### DISCUSSION TITLE
Can I use this project with an AMD GPU?

### DISCUSSION THREAD

**Discussion by theRoboxx** [rakuri255/UltraSinger#78] [Q&A]
Did anyone encounter this problem as well? I cannot install the cuda drivers without a NVIDA GPU, or is there a workaround?
I found [this](https://gist.github.com/Nitesh8998/a337c29a41e23e4c5d0e421852e62db3) article and wanted to ask, if anyone did try it that way?

**Comment by rakuri255** [comment]
There is no cuda for AMD GPUs. AMD working on the ROCm as an alternative. 
But i think its not released yet.

Currently this 3 UltraSinger parts needs cuda because of the Torch lib. Demucs, Whisper and in future crepe. Maybe they have an alternative GPU support? Someone should check it :)

Your article sounds interessting. But what i dont understand is, if he has added/hacked cuda into an AMD GPU or is there already a second GPU that supports cuda?

Maybe we can try to run it in Google Colab. Seems they have also cuda support.

**Reply by BWagener** [reply]
I think it is important to point out this project works perfectly fine without any GPU.
I have had more accurate results (not faster ofc) running it on CPU only on a Steam Deck than running it with my GPU (GTX 1080).

**Reply by theRoboxx** [reply]
What are the commandline parameters for CPU-use only?
I tried it with `py UltraSinger.üy -i "youtube.com/..."` and got the error `no cuda device found` because of my AMD GPU.

**Reply by rakuri255** [reply]
@BWagener Thats interessting. It should be more accurate with an GPU. Can you share the differance?

@theRoboxx When it says there is no cuda than it should use the CPU automaticly. You should also see it in the log for Seperation, Transcription and Pitching -> "... CPU as worker". But i wrote in the ReadMe how to force the CPU. `--force_whisper_cpu` `--force_separation_cpu`

**Reply by BWagener** [reply]
I can't reproduce the problem I was facing on the latest version anymore.

It was an error when pitching the syllables, saying that some syllables pitch was set to 0 due to them being too short or having length 0 I believe. Unfortunately I didn't save the exact error message.
When it was working fine on the Steam Deck I chalked it up to the old GPU, which I guess was not the problem after all.

**Reply by rakuri255** [reply]
@BWagener ah this error. It was a bug that is fixed. See #58
Thanks :)

**Reply by theRoboxx** [reply]
Hi there,
I tried your solution with both flags but now I get an error with demucs.

Did I miss something obvious?
I followed each step in the readme.
```
$ ../.venv/bin/python3.10 UltraSinger.py -i "../us_songs/artist/test_song.mp3" --force_whisper_cpu --force_separation_cpu

2023-09-04 15:49:31.971603: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-09-04 15:49:32.423786: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
[UltraSinger] There are no cuda devices available. -> Using cpu.
[UltraSinger] full automatic mode
[UltraSinger] Searching song in musicbrainz
[UltraSinger] Found Titel and Artist test by Anon
[UltraSinger] Found release year: 2013
[UltraSinger] Found genres: rock,electronic,alternative rock,
[UltraSinger] Creating output folder. -> ../us_songs/artist/output/artist
[UltraSinger] Creating output folder. -> ../us_songs/artist/output/artist/cache
[UltraSinger] Separating vocals from audio with demucs and cpu as worker.
Traceback (most recent call last):
  File "/home/user/Downloads/UltraSinger/src/UltraSinger.py", line 831, in <module>
    main(sys.argv[1:])
  File "/home/user/Downloads/UltraSinger/src/UltraSinger.py", line 757, in main
    run()
  File "/home/user/Downloads/UltraSinger/src/UltraSinger.py", line 303, in run
    audio_separation_path = separate_vocal_from_audio(
  File "/home/user/Downloads/UltraSinger/src/UltraSinger.py", line 432, in separate_vocal_from_audio
    separate_audio(ultrastar_audio_input_path, cache_path, device)
  File "/home/user/Downloads/UltraSinger/src/modules/Audio/separation.py", line 23, in separate_audio
    subprocess.run(
  File "/usr/lib/python3.10/subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.10/subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.10/subprocess.py", line 1863, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'demucs'
```

**Reply by rakuri255** [reply]
@theRoboxx  You dont need `--force_whisper_cpu` `--force_separation_cpu` flags when you dont have a cuda GPU. UltraSinger says that `[UltraSinger] There are no cuda devices available. -> Using cpu.`. So, it uses CPU anyway. 

It shouldn't matter if you use the flags or not. Can you please test this?
I suspect that your `venv` is not allowed to run `demucs`. 
Or the requirement.txt was not installed correctly. 
Can you please try to recreate the `venv`?

**Reply by theRoboxx** [reply]
Unfortunately, the recreation didn't fix anything. Any other tips for finding the bug or my error?

**Reply by rakuri255** [reply]
Try to add `shell=True` in subprocess.
It's in `src/modules/Audio/seperation.py`
Change the line :24 to `["demucs", "-d", device, "--two-stems=vocals", input_file_path], shell=True`

**Comment by rakuri255** [comment]
Maybe try this [coriander](https://github.com/hughperkins/coriander). It compiles cuda applications into openCL. But dont know if there is a benefit in process speed or accuracy.

**Comment by eadmaster** [comment]
Also worth a try: https://github.com/vosen/ZLUDA
